hazelcast/src/main/java/com/hazelcast/cache/impl/TenantContextual.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.cache.impl;
+
+import com.hazelcast.spi.tenantcontrol.TenantControl;
+import com.hazelcast.spi.tenantcontrol.TenantControl.Closeable;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Supplier;
+
+/**
+ * Represents a value that requires tenant control context to be accessed
+ *
+ * @author lprimak
+ * @param <T> object type
+ */
+public class TenantContextual<T> {
```

Comment:
Every problem in computer science can be added by adding an extra layer of indirection; apart from having too many layers of indirection. I have the feeling that we are adding an unwanted layer of indirection here.

Reply:
I am not sure what is the alternative here is, do you have a suggestion?

Reply:
Resolving this comment as this class serves its' purpose and I asked if the tenant context can be resolved eagerly in other comment above.